README.rst
.. These are examples of badges you might want to add to your README:
   please update the URLs accordingly

    .. image:: https://api.cirrus-ci.com/github/<USER>/seren.svg?branch=main
        :alt: Built Status
        :target: https://cirrus-ci.com/github/<USER>/seren
    .. image:: https://readthedocs.org/projects/seren/badge/?version=latest
        :alt: ReadTheDocs
        :target: https://seren.readthedocs.io/en/stable/
    .. image:: https://img.shields.io/coveralls/github/<USER>/seren/main.svg
        :alt: Coveralls
        :target: https://coveralls.io/r/<USER>/seren
    .. image:: https://img.shields.io/pypi/v/seren.svg
        :alt: PyPI-Server
        :target: https://pypi.org/project/seren/
    .. image:: https://img.shields.io/conda/vn/conda-forge/seren.svg
        :alt: Conda-Forge
        :target: https://anaconda.org/conda-forge/seren
    .. image:: https://pepy.tech/badge/seren/month
        :alt: Monthly Downloads
        :target: https://pepy.tech/project/seren
    .. image:: https://img.shields.io/twitter/url/http/shields.io.svg?style=social&label=Twitter
        :alt: Twitter
        :target: https://twitter.com/seren

.. image:: https://img.shields.io/badge/-PyScaffold-005CA0?logo=pyscaffold
    :alt: Project generated with PyScaffold
    :target: https://pyscaffold.org/

|

=====
seren
=====


    Add a short description here!


A longer description of your project goes here...


.. _pyscaffold-notes:

Making Changes & Contributing
=============================

This project uses `pre-commit`_, please make sure to install it before making any
changes::

    pip install pre-commit
    cd seren
    pre-commit install

It is a good idea to update the hooks to the latest version::

    pre-commit autoupdate

Don't forget to tell your contributors to also install and use pre-commit.

.. _pre-commit: https://pre-commit.com/

Note
====

This project has been set up using PyScaffold 4.2.3. For details and usage
information on PyScaffold see https://pyscaffold.org/.


CHANGELOG.md
## 0.3.0 (2022-07-22)

### Fix

- **SystemdNspawn-test**: remove sn reference on systemd-nspawn test

### Refactor

- **Skeleton**: remove skeleton files (#4)
- Rename Systemd-Nspawn module

### Feat

- Add SystemdNspawn module
- created struct of te systemd-nspawn module
- **SystemdNspawn**: Add SystemdNspawn methods, propriety checkers and start run method
- **Finish-SystemdClass**: Finish SystemdClass


tests/conftest.py
"""
    Dummy conftest.py for seren.

    If you don't know what this is for, just leave it empty.
    Read more about conftest.py under:
    - https://docs.pytest.org/en/stable/fixture.html
    - https://docs.pytest.org/en/stable/writing_plugins.html
"""

# import pytest


tests/test_mount.py
import os
from multiprocessing import Queue

import pytest

from seren.mount import FilesSystemType as FS
from seren.mount import Mount
from seren.mount import MountFlags as MFLAGS
from seren.mount import OverlayFs

__author__ = "Odair M."
__copyright__ = "Odair M."
__license__ = "MIT"


@pytest.mark.parametrize(
    "test_input,expected",
    [
        ("EXT4", "ext4"),
        ("TMPFS", "tmpfs"),
        ("DEVPTS", "devpts"),
        ("OVERLAYFS", "overlay"),
        pytest.param("DEV", "dev", marks=pytest.mark.xfail),
        pytest.param("fat", "", marks=pytest.mark.xfail),
    ],
)
def test_file_system_type(test_input: str, expected: str):
    """Test filesystem type
    This class map filesystem typefor constants like enumeration
    :param test_input: A string that represent test
    :param expected: expected response

    """
    assert FS[test_input].value == expected


@pytest.mark.parametrize(
    "test_input,expected",
    [
        ("MS_RDONLY", 0),
        ("MS_BIND", 2**12),
        ("MS_PRIVATE", 18),
        pytest.param("DEV", "dev", marks=pytest.mark.xfail),
        pytest.param("fat", -1, marks=pytest.mark.xfail),
    ],
)
def test_mount_flags(test_input: str, expected: int):
    """Test if the class MountFlags returns corrects values

    :param test_input: enumarate name
    :param expected: int number

    """
    assert MFLAGS[test_input].value == expected


@pytest.fixture
def queue():
    """Return a multiprocessing queue fixture
    :returns: MultiProcessing queue

    """
    q = Queue()
    yield q
    q.close()


def test_mount_targets(tmp_path):
    """Test if the targest mounts is the same

    :param tmp_path: TODO
    :returns: TODO

    """
    source = "/dev/pts"
    target = str(tmp_path.absolute())

    devpts = Mount(source, target, FS.DEV, MFLAGS.MS_BIND)

    assert target == devpts.get_target()


def test_mount_delete(tmp_path):
    """Test umount feature for the class

    :param tmp_path: TODO
    :returns: TODO

    """
    source = "/sys"
    target = str(tmp_path.absolute())
    sysfs = Mount(source, target, FS.SYSFS, MFLAGS.MS_BIND)
    assert os.path.ismount(target)
    del sysfs
    assert os.path.ismount(target) is False


def test_mount_empty_target():
    """TODO: Docstring for test_mount_empty_target.

    :returns: TODO

    """
    with pytest.raises(OSError):
        t = Mount("/proc", "", FS.PROC, MFLAGS.MS_RDONLY)
        del t


def test_overlay(tmp_path_factory):
    """Test overlay class, mount overlayfs and check if is ok

    :param tmp_path_factory: TODO
    :returns: TODO

    """
    base_dir = str(tmp_path_factory.mktemp("base").absolute())
    work_dir = str(tmp_path_factory.mktemp("work").absolute())
    lower_dir = str(tmp_path_factory.mktemp("lower1").absolute())
    upperdir = str(tmp_path_factory.mktemp("upper").absolute())
    target = str(tmp_path_factory.mktemp("target").absolute())

    overlay = OverlayFs(base_dir, work_dir, upperdir, target, lower_dir)

    # time.sleep(60)
    assert os.path.ismount(overlay.get_target())


def test_mount_proc(tmp_path):
    """Test if the mount class can bind /proc

    :param tmp_path: fixture for temporary directory

    """
    source = "/proc"
    target = str(tmp_path.absolute())

    proc = Mount(source, target, FS.PROC, MFLAGS.MS_BIND)

    assert os.path.ismount(proc.get_target())


tests/test_skeleton.py
import pytest

from seren.skeleton import fib, main

__author__ = "Odair M."
__copyright__ = "Odair M."
__license__ = "MIT"


def test_fib():
    """API Tests"""
    assert fib(1) == 1
    assert fib(2) == 1
    assert fib(7) == 13
    with pytest.raises(AssertionError):
        fib(-10)


def test_main(capsys):
    """CLI Tests"""
    # capsys is a pytest fixture that allows asserts against stdout/stderr
    # https://docs.pytest.org/en/stable/capture.html
    main(["7"])
    captured = capsys.readouterr()
    assert "The 7-th Fibonacci number is 13" in captured.out


src/seren/settings.py
"""
File: settings.py
Author: Odair M.
Email: odairmario45@gmail.com
Github: https://github.com/odairmario/seren
Description: Settings from the project
"""
import logging

LICENSE = "MIT"
AUTHOR = "Odair M."
COPYRIGHT = "Odair M."
GITHUB = "https://github.com/odairmario/seren"

logs_level = "DEBUG"
# logs_directory = logs_conf.get("directory","")
# logs_output = os.path.join(logs_directory,logs_conf.get("path","experiment.log"))
# logs_output_rotation = os.path.join(
#        logs_directory,
#        logs_conf.get("file_rotation","experiment.old.log")
#        )
# logs_handles = [
#    FileHandler(
#        filename=logs_output,
#        encoding='utf-8'
#    ),
#    RotatingFileHandler(
#    logs_output_rotation,
#    backupCount=logs_conf.get("backups_count",5),
#    maxBytes=logs_conf.get("max_size",2**20),
#    encoding="utf-8"
#    ),
#    StreamHandler()
# ]

logging.basicConfig(
    encoding="utf-8",
    format="%(asctime)s:%(levelname)s:%(message)s",
    level=logging.getLevelName(logs_level),
    # handlers=logs_handles
)


src/seren/skeleton.py
"""
This is a skeleton file that can serve as a starting point for a Python
console script. To run this script uncomment the following lines in the
``[options.entry_points]`` section in ``setup.cfg``::

    console_scripts =
         fibonacci = seren.skeleton:run

Then run ``pip install .`` (or ``pip install -e .`` for editable mode)
which will install the command ``fibonacci`` inside your current environment.

Besides console scripts, the header (i.e. until ``_logger``...) of this file can
also be used as template for Python modules.

Note:
    This file can be renamed depending on your needs or safely removed if not needed.

References:
    - https://setuptools.pypa.io/en/latest/userguide/entry_point.html
    - https://pip.pypa.io/en/stable/reference/pip_install
"""

import argparse
import logging
import sys

from seren import __version__

__author__ = "Odair M."
__copyright__ = "Odair M."
__license__ = "MIT"

_logger = logging.getLogger(__name__)


# ---- Python API ----
# The functions defined in this section can be imported by users in their
# Python scripts/interactive interpreter, e.g. via
# `from seren.skeleton import fib`,
# when using this Python module as a library.


def fib(n):
    """Fibonacci example function

    Args:
      n (int): integer

    Returns:
      int: n-th Fibonacci number
    """
    assert n > 0
    a, b = 1, 1
    for _i in range(n - 1):
        a, b = b, a + b
    return a


# ---- CLI ----
# The functions defined in this section are wrappers around the main Python
# API allowing them to be called directly from the terminal as a CLI
# executable/script.


def parse_args(args):
    """Parse command line parameters

    Args:
      args (List[str]): command line parameters as list of strings
          (for example  ``["--help"]``).

    Returns:
      :obj:`argparse.Namespace`: command line parameters namespace
    """
    parser = argparse.ArgumentParser(description="Just a Fibonacci demonstration")
    parser.add_argument(
        "--version",
        action="version",
        version="seren {ver}".format(ver=__version__),
    )
    parser.add_argument(dest="n", help="n-th Fibonacci number", type=int, metavar="INT")
    parser.add_argument(
        "-v",
        "--verbose",
        dest="loglevel",
        help="set loglevel to INFO",
        action="store_const",
        const=logging.INFO,
    )
    parser.add_argument(
        "-vv",
        "--very-verbose",
        dest="loglevel",
        help="set loglevel to DEBUG",
        action="store_const",
        const=logging.DEBUG,
    )
    return parser.parse_args(args)


def setup_logging(loglevel):
    """Setup basic logging

    Args:
      loglevel (int): minimum loglevel for emitting messages
    """
    logformat = "[%(asctime)s] %(levelname)s:%(name)s:%(message)s"
    logging.basicConfig(
        level=loglevel, stream=sys.stdout, format=logformat, datefmt="%Y-%m-%d %H:%M:%S"
    )


def main(args):
    """Wrapper allowing :func:`fib` to be called with string arguments in a CLI fashion

    Instead of returning the value from :func:`fib`, it prints the result to the
    ``stdout`` in a nicely formatted message.

    Args:
      args (List[str]): command line parameters as list of strings
          (for example  ``["--verbose", "42"]``).
    """
    args = parse_args(args)
    setup_logging(args.loglevel)
    _logger.debug("Starting crazy calculations...")
    print("The {}-th Fibonacci number is {}".format(args.n, fib(args.n)))
    _logger.info("Script ends here")


def run():
    """Calls :func:`main` passing the CLI arguments extracted from :obj:`sys.argv`

    This function can be used as entry point to create console scripts with setuptools.
    """
    main(sys.argv[1:])


if __name__ == "__main__":
    # ^  This is a guard statement that will prevent the following code from
    #    being executed in the case someone imports this file instead of
    #    executing it as a script.
    #    https://docs.python.org/3/library/__main__.html

    # After installing your project with pip, users can also run your Python
    # modules as scripts via the ``-m`` flag, as defined in PEP 338::
    #
    #     python -m seren.skeleton 42
    #
    run()


src/seren/systemdNspawn.py
"""
File: systemd-nspawn.py
Author: Odair M.
Email: odairmario45@gmail.com
Github: https://github.com/odairmario/seren
Description: Systemd-nspawn module
"""

import shutil
import subprocess as sp
from os import path

__author__ = "Odair M."
__copyright__ = "Odair M."
__license__ = "MIT"


class SystemdNspawn(object):

    """SystemdNspawn represent a wrap to call systemd-nspawn."""

    def __init__(self, target_value, **kwargs):
        """constructor of the class

        :param target: mount point

        """
        self.target = target_value
        self.systemd_nspawn = kwargs.get("nspawn_path", "systemd-nspawn")
        self.nspawn_arguments = kwargs.get(
            "nspawn_arguments", ["-q -D {}".format(self.target)]
        )

        # Stdout and Stderr
        self._stderr = None
        self._stdout = None

    @property
    def nspawn_arguments(self):
        """Systemd-nspawn arguments
        :returns: TODO

        """

        return self._nspawn_arguments

    @nspawn_arguments.setter
    def nspawn_arguments(self, arguments: list):
        """Setter for nspawn_arguments, must be a list

        :arguments: TODO
        :returns: TODO

        """

        if type(arguments) is list:
            self._nspawn_arguments = arguments
        else:
            raise ValueError("Arguments is not a list!")

    @property
    def target(self):
        """Target getter
        :returns: target path a string

        """

        return self._target

    @target.setter
    def target(self, target_value):
        """Setter for target atribute

        :target_value: TODO
        :returns: TODO

        """

        if path.isdir(target_value):
            self._target = target_value
        else:
            raise ValueError(
                "Target path %s dosen't exist or not is a directory.", target_value
            )

    @property
    def systemd_nspawn(self):
        """systemd_nspawn getter
        :returns: TODO

        """

        return self._systemd_nspawn

    @systemd_nspawn.setter
    def systemd_nspawn(self, nspawn):
        """TODO: Docstring for systemd_nspawn.

        :nspawn: TODO
        :returns: TODO

        """

        if shutil.which(nspawn):
            self._systemd_nspawn = nspawn
        else:
            raise Exception("systemd-nspawn not found")

    def get_last_output(self):
        """Return last output string
        :returns: TODO

        """

        return self._stdout

    def __enter__(self):
        """__enter__ method for WITH stament

        :param type: TODO
        :param value: TODO
        :param traceback: TODO
        :returns: TODO

        """

        return self

    # TODO
    def __exit__(self, type, value, traceback):
        """Exit __exit____exi____exit__
        :returns: TODO

        """

    def run(self, *commands):
        """Run a command on namespace container

        :param *commands: TODO
        :returns: TODO

        """
        command = " ".join([self.systemd_nspawn, *self.nspawn_arguments, *commands])

        print(command)
        try:
            err = sp.run(
                command, shell=True, capture_output=True, check=True, text=True
            )
            self._stdout = err.stdout
            self._stderr = err.stderr

            return err.returncode
        except Exception as e:
            raise e


src/seren/systemd-nspawn.py
"""
File: systemd-nspawn.py
Author: Odair M.
Email: odairmario45@gmail.com
Github: https://github.com/odairmario/seren
Description: Systemd-nspawn module
"""

# from subprocess import Popen

__author__ = "Odair M."
__copyright__ = "Odair M."
__license__ = "MIT"


class SystemdNspawn(object):

    """SystemdNspawn represent a wrap to call systemd-nspawn."""

    def __init__(self, target, **kwargs):
        """constructor of the class

        :param target: mount point

        """
        self._target = target

    def __enter__(self, type, value, traceback):
        """__enter__ method for WITH stament

        :param type: TODO
        :param value: TODO
        :param traceback: TODO
        :returns: TODO

        """

        return self

    # TODO
    def __exit__(self):
        """Exit __exit____exi____exit__
        :returns: TODO

        """

    def run(self, *commands):
        """Run a command on namespace container

        :param *commands: TODO
        :returns: TODO

        """


src/seren/mount.py
"""
File: mount.py
Author: Odair M.
Email: odairmario45@gmail.com
Github: https://github.com/odairmario
Description: Mount module
"""
import ctypes
import ctypes.util
import logging as log
import os
from enum import Enum

from seren.settings import AUTHOR, COPYRIGHT, LICENSE

__author__ = AUTHOR
__copyright__ = COPYRIGHT
__license__ = LICENSE


class NoValue(Enum):
    def __repr__(self):
        return "<%s.%s>" % (self.__class__.__name__, self.name)


class FilesSystemType(NoValue):
    """Filesystem types map.
    This class map filesystem typefor constants like enumeration
    """

    TMPFS = "tmpfs"
    EXT4 = "ext4"
    OVERLAYFS = "overlay"
    PROC = "proc"
    DEV = "devtmpfs"
    SYSFS = "sysfs"
    DEVPTS = "devpts"
    EMPTY = ""


class MountFlags(Enum):
    """Mount flags for the *mount* syscall"""

    MS_RDONLY = 0
    MS_NOSUID = 1
    MS_NODEV = 2
    MS_NOEXEC = 3
    MS_SYNCHRONOUS = 4
    MS_REMOUNT = 5
    MS_MANDLOCK = 6
    MS_DIRSYNC = 7
    MS_NOATIME = 10
    MS_NODIRATIME = 11
    MS_BIND = 4096
    MS_BIND_PRIVATE = 266240
    MS_MOVE = 13
    MS_REC = 14
    MS_SILENT = 15
    MS_POSIXACL = 16
    MS_UNBINDABLE = 17
    MS_PRIVATE = 18
    MS_SLAVE = 19
    MS_SHARED = 20
    MS_RELATIME = 21
    MS_KERNMOUNT = 22
    MS_I_VERSION = 23
    MS_STRICTATIME = 24
    MS_ACTIVE = 30
    MS_NOUSER = 31


def umount(target: str):
    """Ummount filesystem

    :param target: mount point of the file system
    :returns: TODO

    """

    if os.path.ismount(target) is False:
        log.warning("Filesystem %s is not mounted. Nothing to do!", target)

        return False

    libc = ctypes.CDLL(ctypes.util.find_library("c"), use_errno=True)
    # ret_code = libc.umount(ctypes.c_char_p(target.encode()))
    ret_code = libc.umount2(ctypes.c_char_p(target.encode()), ctypes.c_ulong(1))

    if ret_code < 0:
        log.error(
            "Error on umount the %s. Error code: %s %s",
            target,
            ret_code,
            ctypes.get_errno(),
        )
        # libc.umount2(ctypes.c_char_p(target.encode()),ctypes.c_ulong(0))
        raise OSError(ctypes.get_errno())

        return False

    if os.path.ismount(target) is True:
        log.error("Filesystem %s still mounted", target)
        # raise Exception("Filesystem still mounted %s",target)

        return False

    return True


class Mount(object):

    """Class to represent mount syscall.
    When instantiate the class mount is do and when is destroyed
    the mount point is umounted.
    """

    def __init__(self, source, target, fs, flag, **data):
        """Call mount syscall and mount the filesystem.
            Mount a filesystem with parameters given, raise error if could not mount
        .

            :param source: source of the filesystem
            :param target: directory that the filesystem will be mounted
            :param fs: filesystem type
            :param flags: list of mount flags
            :param **data: additional data passed for the syscall.

        """
        self._source = source
        self._target = target
        self._fs = fs.value
        self._flag = flag.value
        self._data = ",".join(["{}={}".format(k, v) for k, v in data.items()])

        self._ret_code = None  # define syscall return code to None

        # define libc calls
        self.__libc = ctypes.CDLL(ctypes.util.find_library("c"), use_errno=True)
        self.__libc.mount.argtypes = (
            ctypes.c_char_p,
            ctypes.c_char_p,
            ctypes.c_char_p,
            ctypes.c_ulong,
            ctypes.c_char_p,
        )
        self._ret_code = self.__libc.mount(
            self._source.encode(),
            self._target.encode(),
            self._fs.encode(),
            self._flag,
            self._data.encode(),
        )

        if self._ret_code < 0:
            self._errno = ctypes.get_errno()
            raise OSError(
                self._errno,
                "Error mounting {} ({} on {}) with options {}: {}".format(
                    self._source,
                    self._fs,
                    self._target,
                    self._data,
                    os.strerror(self._errno),
                ),
            )

    def get_target(self):
        """Return the mountpoint where the filesystem are mounted.
        :returns: string

        """

        if os.path.ismount(self._target):
            return self._target
        else:
            return ""

    def __del__(self):
        """Destroy object, umount fs
        :returns: TODO

        """
        umount(self._target)

    def __enter__(self, type, value, traceback):
        """__enter method for WiTH stament
        :returns: TODO

        """

        return self

    def __exit__(self):
        """Exit stament__exit____exit__x
        :returns: TODO

        """
        pass
        # umount(self._target)


class OverlayFs(Mount):

    """Mount overlayfs system"""

    __SOURCE = "overlay"

    def __init__(
        self, base_dir: str, work_dir: str, upperdir: str, target: str, *lowers_dir
    ):
        """Mount a overlayfs filesystem

        mount -t overlay overlay -o lowerdir=l2:l1,upperdir=upper,workdir=workdir merged
        :param base_dir: TODO
        :param work_dir: TODO
        :param upperdir: TODO
        :param target: TODO
        :param *lowers_dir: TODO

        """
        self._base_dir = base_dir
        self._work_dir = work_dir
        self._upperdir = upperdir
        self._target = target
        self._lowers = lowers_dir
        data = {
            "lowerdir": ":".join([*self._lowers, self._base_dir]),
            "workdir": self._work_dir,
            "upperdir": self._upperdir,
        }
        Mount.__init__(
            self,
            self.__SOURCE,
            self._target,
            FilesSystemType.OVERLAYFS,
            MountFlags.MS_ACTIVE,
            **data
        )
